api_app/helpers.py

```diff
@@ -130,3 +133,43 @@ def get_hash_type(hash_value):
             detected_hash_type = hash_type
             break
     return detected_hash_type  # stays None if no matches
+
+
+def get_custom_config_as_dict(user, entity_type, name=None) -> dict:
+    """
+    Returns custom config as dict
+    """
+    result = {}
+    try:
+        membership = Membership.objects.get(user=user)
+        custom_configs = CustomConfig.objects.filter(
+            organization=membership.organization,
+            type=entity_type,
+        )
+        if name is not None:
+            custom_configs = custom_configs.filter(name=name)
+        for custom_config in custom_configs:
+            custom_config: CustomConfig
+            if custom_config.name not in result:
+                result[custom_config.name] = {}
+            result[custom_config.name][custom_config.attribute] = json.loads(
```

Comment:
the `json.loads` call sounds wrong cause this means that we are erronously saving the `value` attribute as valid JSON while it should just be a simple string

Reply:
The json conversion is done as the param value it is overriding is guaranteed to be a JSON-compliant value. This allows us to store all the JSON-compliant types in the StringField

Reply:
Even if JSON is a string, StringField/CharField is not made to store JSON data. 
Anyway I don't even understand why you are using JSON instead of simple string for "value".

We have a structure like this:
```
 "Classic_DNS": {
    "type": "observable",
    "python_module": "dns.dns_resolvers.classic_dns_resolver.ClassicDNSResolver",
    "description": "Retrieve current domain resolution with default DNS",
    "disabled": false,
    "external_service": true,
    "leaks_info": false,
    "observable_supported": ["ip", "domain", "url"],
    "config": {
      "soft_time_limit": 30,
      "queue": "default"
    },
    "secrets": {},
    "params": {
      "query_type": {
        "value": "A",
        "type": "str",
        "description": "Query type against the chosen DNS resolver."
      }
    }
  }
```

The "value" attribute in CustomConfig should just be a simple string "A" in this case or "CNAME" in the example that you made.

Reply:
you got some problems on how you handle that parameter:

looking at what you have written, why CNAME is escaped and containers a string inside a string?
```
    "value": "\"CNAME\"",

```

this should just be
```
 "value": "CNAME",
```

Reply:
Let's say a param has type str and the value in it is `5`. Now, we'll store `"5"` in CustomConfig.value. However, if it is of type int and the value in it is `5`, we'll store `5`

Reply:
So, when I json.loads it, it will return the correct data type

Reply:
Ok now it starts to make sense why you did that but:
1 - that was completely implicit
2 - if we need to make data conversion, we could do that in another way. Here we are abusing JSON to get the result that we want. It is not meant to be used to do such kind of things. This is an hack we should avoid.

I think we should leverage what was already built inside IntelOwl to manage this properly. De facto what we want is to store a new configuration for a specific parameter in the analyzer config file. But we already have a Serializer for that config file (AbstractConfigSerializer).
If you check [here](https://github.com/intelowlproject/IntelOwl/blob/master/api_app/core/serializers.py#L95), there is already a `ParamSerializer` used for that structure that we want to override.

The right way to do that could be: first accept a value from the API. Verify that this value is complaint with the existing structure (if `type` is `str` it must be a string) and then save it as a BaseField (see [here](https://github.com/intelowlproject/IntelOwl/blob/master/api_app/core/serializers.py#L51)).
Then we wouldn't ever have any other problems with that parameter cause we can overwrite it here and there securely because we already managed the verification in the serializer properly so the data is always valid.

Reply:
Let's say we receive a `5` as a value by a `form-data` (or `x-www-form-payload type request. Unlike a `JSON` payload, there is no type information in a `form-data`, it's always a string.
![image](https://user-images.githubusercontent.com/58381523/180011610-2cf25130-2d1f-48a0-9a7c-836c5776921b.png)
Hence, value_type will always be str.

So now, we need to get the type of this analyzer attribute from the config file, then cast the payload to the said format and handle errors. Moreover, if (in the future) we want to support object-type values, we'll need to stringify them. 

However, considering that the `analyzer-config` file itself is a JSON file, it is a guarantee that the JSON-parsing mechanism will always continue to work.

Reply:
I am not getting why we should received a `form-data` payload in this API. All the other endpoints work by sending and receiving JSON. `Content-Type` of both answers and requests is always `application/json`.

I mean, you wrote this example:
```
POST request at /api/custom-config with body:

{
    "name":"Classic_DNS",
    "attribute":"query_type",
    "value":"\"CNAME\"",
    "type": 1
}
```

Where the payload is clearly JSON but the `value` is JSON itself. That's what I don't understand

Reply:
Yeah, you're right. Currently, we're using `application/json` everywhere (except file uploads). But, in the future, if we want to mix this functionality with a file upload somehow, we would be unable to. So, I just wanted to be able to support it if needed in future.
Additionally, (for param values of types like `array` or `object`), we will need to stringify them in future. Thus, those doors are kept open too if we use JSON

Reply:
thanks for the explanation. 

I don't think that this specific API endpoint would ever support that. And even if that will happen, we'll make changes then. Right now I think it is better to keep things simple there.

Then the point regarding having array types that is true. We already have `type:list` params. I don't think we'll ever have `type:dict` because in that case it would make sense to just create another simple param.
Anyway, this was already handled previously by @eshaan7 when writing the code of the `AbstractConfigSerializer` that I linked you before. He leveraged the `BaseField` that basically allow to write every kind of type inside of it.
So, if the request payload is valid JSON and the analyzer config is valid JSON, then why both cases shouldn't be handled the same way?

In `CustomConfig`,
`type` could be `rfs.ChoiceField(choices=PARAM_DATATYPE_CHOICES)` and
`value` could be `BaseField()`

like it is in the `_ParamSerializer` serializer

Reply:
BaseField is not a valid db type. That's what I meant by "we'll need to stringify"... and destringify i.e. `json.dumps` and `json.loads`.

Reply:
ah you are right, damn the difference is that the analyzer config is not in the real DB so that association I pointed out is not possible.

ok so another option that could be more clear (I would like to get rid of that explicit json conversions if possible) is to use the Django [JSONField](https://docs.djangoproject.com/en/4.0/ref/models/fields/#jsonfield). What do you think?

Reply:
I could do that. But databases are widely varying in how they handle JSON Fields, and the explicit conversions will still remain (for type-checking purposes).

Reply:
yeah but if we are saving this as a valid JSON I think it makes sense to explicitly use this field considering that it is supported

Reply:
Alright. So everything else remains same, just the StringField changes to a JSONField, right?